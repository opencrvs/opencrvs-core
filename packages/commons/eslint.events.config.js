/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 *
 * OpenCRVS is also distributed under the terms of the Civil Registration
 * & Healthcare Disclaimer located at http://opencrvs.org/license.
 *
 * Copyright (C) The OpenCRVS Authors located at https://github.com/opencrvs/opencrvs-core/blob/master/AUTHORS.
 */

const eventsConfig = require('../../eslint.events.config.js')
const { defineConfig } = require('eslint/config')
const path = require('path')

module.exports = defineConfig([
  {
    languageOptions: {
      sourceType: 'commonjs',
      parserOptions: {
        project: [path.resolve(__dirname, './tsconfig.json')]
      }
    },
    plugins: {
      import: require('eslint-plugin-import')
    },
    settings: {
      'import/resolver': {
        typescript: {
          project: path.resolve(__dirname, './tsconfig.json')
        }
      }
    },
    files: [
      './src/events/**/*.ts',
      'src/events/**/*.ts',
      'src/conditionals/**/*.ts',
      './src/conditionals/**/*.ts'
    ],
    rules: {
      ...eventsConfig.rules,
      'import/no-relative-parent-imports': 'off',
      '@typescript-eslint/consistent-type-definitions': ['error', 'type'],
      'no-restricted-imports': [
        'error',
        {
          paths: [
            {
              name: '../events',
              message:
                "Don't import from '../events' directly under commons. Use the actual file (e.g., 'events/EventConfig') to avoid Zod type issues."
            },
            {
              name: './events',
              message:
                "Don't import from './events' directly under commons. Use the actual file (e.g., 'events/EventConfig') to avoid Zod type issues."
            }
          ],
          patterns: [
            '**/events/index',
            '**/events/index.ts',
            '**/events/index.js'
          ]
        }
      ]
    }
  }
])
