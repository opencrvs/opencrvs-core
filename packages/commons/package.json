{
  "name": "@opencrvs/commons",
  "version": "1.8.1",
  "description": "OpenCRVS common modules and utils",
  "license": "MPL-2.0",
  "main": "./build/dist/common/index.js",
  "exports": {
    ".": "./build/dist/common/index.js",
    "./monitoring": "./build/dist/common/monitoring.js",
    "./types": "./build/dist/common/types.js",
    "./logger": "./build/dist/common/logger.js",
    "./authentication": "./build/dist/common/authentication.js",
    "./http": "./build/dist/common/http.js",
    "./fixtures": "./build/dist/common/fixtures.js",
    "./assignment": "./build/dist/common/assignment.js",
    "./client": "./build/dist/esm/client.js",
    "./events": "./build/dist/common/events/index.js"
  },
  "scripts": {
    "start": "pnpm build:watch",
    "test": "pnpm test:compilation && jest",
    "test:watch": "jest --watch",
    "test:compilation": "tsc --noEmit",
    "lint": "eslint -c eslint.config.js --fix --max-warnings=0",
    "postinstall": "pnpm build",
    "build:default": "tsc --project tsconfig.build.json",
    "build:esm": "tsc --project tsconfig.esm.build.json",
    "build": "pnpm build:default && pnpm build:esm",
    "build:watch": "concurrently \"tsc --watch --preserveWatchOutput --project tsconfig.build.json\" \"tsc --watch --preserveWatchOutput --project tsconfig.esm.build.json\"",
    "build:clean": "rm -rf build"
  },
  "private": true,
  "dependencies": {
    "@types/fhir": "^0.0.37",
    "@types/hapi__hapi": "^20.0.0",
    "@types/jest": "^26.0.14",
    "@types/jwt-decode": "^2.2.1",
    "@types/lodash": "4.17.0",
    "@types/uuid": "^9.0.3",
    "ajv": "^8.17.1",
    "ajv-formats": "^3.0.1",
    "concurrently": "^9.1.0",
    "date-fns": "^2.28.0",
    "elastic-apm-node": "^3.29.0",
    "jest": "29.7.0",
    "jwt-decode": "^3.0.0",
    "lodash": "^4.17.10",
    "pino": "^7.0.0",
    "pkg-up": "^3.1.0",
    "typescript": "5.6.3",
    "uuid": "^9.0.0",
    "yaml": "^2.7.1",
    "zod": "^3.23.8",
    "zod-openapi": "^4.2.4"
  },
  "devDependencies": {
    "@types/jsonwebtoken": "^9.0.9",
    "@typescript-eslint/eslint-plugin": "8.28.0",
    "@typescript-eslint/parser": "^8.28.0",
    "eslint": "^9.23.0",
    "eslint-config-prettier": "^10.1.1",
    "eslint-import-resolver-typescript": "^4.3.1",
    "eslint-plugin-import": "^2.31.0",
    "eslint-plugin-prettier": "^5.2.5",
    "jest-fetch-mock": "^2.1.2",
    "jsonwebtoken": "^9.0.2",
    "pino-pretty": "^11.0.0",
    "ts-jest": "29.2.5",
    "tsx": "^4.19.4"
  },
  "lint-staged": {
    "src/**/*.ts": [
      "prettier --write",
      "eslint --fix"
    ]
  },
  "jest": {
    "transform": {
      "^.+\\.tsx?$": "ts-jest"
    },
    "coverageReporters": [
      "json",
      "lcov",
      "text",
      "html"
    ],
    "testRegex": "(/__tests__/.*|(\\.|/)(test|spec))\\.tsx?$",
    "setupFiles": [
      "./test/setupJest.ts"
    ],
    "moduleFileExtensions": [
      "ts",
      "tsx",
      "js",
      "jsx",
      "json",
      "node"
    ],
    "testEnvironment": "node"
  }
}
