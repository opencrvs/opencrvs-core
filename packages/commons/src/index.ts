/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 *
 * OpenCRVS is also distributed under the terms of the Civil Registration
 * & Healthcare Disclaimer located at http://opencrvs.org/license.
 *
 * Copyright (C) The OpenCRVS Authors located at https://github.com/opencrvs/opencrvs-core/blob/master/AUTHORS.
 */
export * from './token-verifier'
export * from './uuid'
export * from './documents'
export * from './http'
export * from './logger'
export * from './roles'
export * from './search'
export * from './events'
export * from './users/service'
export * from './authentication'
