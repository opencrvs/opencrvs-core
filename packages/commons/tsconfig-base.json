{
  "compilerOptions": {
    "baseUrl": ".",
    "paths": {
      "src/*": [
        "./src/*"
      ]
    },
    "target": "es6",
    "sourceMap": true,
    "rootDir": "src",
    "declaration": true,
    "lib": [
      "esnext.asynciterable",
      "es2015",
      "es6",
      "es2019"
    ],
    "forceConsistentCasingInFileNames": true,
    "noImplicitReturns": true,
    "noImplicitThis": true,
    "noImplicitAny": true,
    "strictNullChecks": true,
    "noUnusedLocals": true,
    "types": [
      "jest",
      "fhir"
    ],
    "skipLibCheck": true,
    "declarationMap": true,
    "composite": true
  }
}
