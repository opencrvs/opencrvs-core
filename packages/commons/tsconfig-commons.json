{
  "extends": "./tsconfig-base.json",
  "compilerOptions": {
    "module": "commonjs",
    "outDir": "build/dist/common"
  },
  "references": [],
  "include": [
    "src/**/*.ts",
    "typings"
  ],
  "exclude": [
    "src/**/*.test.ts"
  ]
}
