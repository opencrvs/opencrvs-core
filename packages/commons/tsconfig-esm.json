{
  "extends": "./tsconfig-base.json",
  "compilerOptions": {
    "module": "esnext",
    "moduleResolution": "node",
    "outDir": "build/dist/esm",
  },
  "references": [],
  "include": [
    "src/**/*.ts"
  ],
  "exclude": [
    "src/**/*.test.ts",
    "src/monitoring.ts"
  ]
}
