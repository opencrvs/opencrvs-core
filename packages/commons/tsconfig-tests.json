{
  "extends": "./tsconfig-base.json",
  "compilerOptions": {
    "module": "commonjs"
  },
  "references": [
    {
      "path": "./tsconfig-commonjs.json"
    }
  ],
  "include": [
    "src/**/*.test.ts",
    "typings"
  ]
}
