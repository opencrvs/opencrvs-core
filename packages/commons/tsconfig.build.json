{
  "extends": "./tsconfig.json",
  "exclude": [
    "node_modules",
    "build",
    "test/setupJest.ts",
    "src/client.ts",
    "src/**/*.test.ts"
  ]
}
