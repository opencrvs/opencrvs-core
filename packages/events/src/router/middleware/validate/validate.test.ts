/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 *
 * OpenCRVS is also distributed under the terms of the Civil Registration
 * & Healthcare Disclaimer located at http://opencrvs.org/license.
 *
 * Copyright (C) The OpenCRVS Authors located at https://github.com/opencrvs/opencrvs-core/blob/master/AUTHORS.
 */
import { ConditionalType, field, FieldType } from '@opencrvs/commons'
import { getFieldErrors } from './index'

describe('getFieldErrors()', () => {
  it('should return an empty array there are no fields to validate', () => {
    const errors = getFieldErrors([], {}, {})
    expect(errors).toEqual([])
  })

  it('should return an error if a required field is not provided', () => {
    const errors = getFieldErrors(
      [
        {
          id: 'test.checkbox',
          type: FieldType.CHECKBOX,
          required: true,
          label: {
            id: 'test.checkbox.label',
            defaultMessage: 'Test Field',
            description: 'Test Field Description'
          }
        }
      ],
      {},
      {}
    )
    expect(errors).toMatchSnapshot()
  })

  it('should not return an error if a required field is provided', () => {
    const errors = getFieldErrors(
      [
        {
          id: 'test.checkbox',
          type: FieldType.CHECKBOX,
          required: true,
          label: {
            id: 'test.checkbox.label',
            defaultMessage: 'Test Field',
            description: 'Test Field Description'
          }
        }
      ],
      { 'test.checkbox': true }
    )

    expect(errors).toMatchSnapshot()
  })

  it('should return an error if a value for a conditionally hidden required field is provided', () => {
    const errors = getFieldErrors(
      [
        {
          id: 'test.checkbox',
          type: FieldType.CHECKBOX,
          required: true,
          label: {
            id: 'test.field.label',
            defaultMessage: 'Test Field',
            description: 'Test Field Description'
          },
          conditionals: [
            {
              type: ConditionalType.SHOW,
              conditional: field('test.text').isEqualTo('helloooo')
            }
          ]
        }
      ],
      { 'test.checkbox': true },
      {}
    )

    expect(errors).toMatchSnapshot()
  })

  it('should not return an error if a value for a conditionally hidden required field is not provided', () => {
    const errors = getFieldErrors(
      [
        {
          id: 'test.checkbox',
          type: FieldType.CHECKBOX,
          required: true,
          label: {
            id: 'test.field.label',
            defaultMessage: 'Test Field',
            description: 'Test Field Description'
          },
          conditionals: [
            {
              type: ConditionalType.SHOW,
              conditional: field('test.text').isEqualTo('helloooo')
            }
          ]
        }
      ],
      {},
      {}
    )

    expect(errors).toMatchSnapshot()
  })

  it('should not return an error if a value for a conditionally visible required field is provided', () => {
    const errors = getFieldErrors(
      [
        {
          id: 'test.checkbox',
          type: FieldType.CHECKBOX,
          required: true,
          label: {
            id: 'test.field.label',
            defaultMessage: 'Test Field',
            description: 'Test Field Description'
          },
          conditionals: [
            {
              type: ConditionalType.SHOW,
              conditional: field('test.text').isEqualTo('helloooo')
            }
          ]
        }
      ],
      { 'test.checkbox': true },
      { 'test.text': 'helloooo' }
    )

    expect(errors).toMatchSnapshot()
  })

  it('should not return error if multiple fields with same id are configured, and value is provided', () => {
    const errors = getFieldErrors(
      [
        {
          id: 'test.input',
          type: FieldType.TEXT,
          required: true,
          label: {
            id: 'test.field.label',
            defaultMessage: 'Test Field',
            description: 'Test Field Description'
          },
          conditionals: [
            {
              type: ConditionalType.SHOW,
              conditional: field('test.text').isEqualTo('helloooo')
            }
          ]
        },
        {
          id: 'test.input',
          type: FieldType.TEXT,
          required: true,
          label: {
            id: 'test.field.label',
            defaultMessage: 'Test Field',
            description: 'Test Field Description'
          },
          conditionals: [
            {
              type: ConditionalType.SHOW,
              conditional: field('test.other').isEqualTo('helloooo')
            }
          ]
        }
      ],
      { 'test.checkbox': true },
      { 'test.text': 'helloooo' }
    )

    expect(errors).toMatchSnapshot()
  })

  it('should return errors for field with custom validation if value does not pass validation', () => {
    const errors = getFieldErrors(
      [
        {
          id: 'test.input',
          type: FieldType.TEXT,
          required: true,
          label: {
            id: 'test.field.label',
            defaultMessage: 'Test Field',
            description: 'Test Field Description'
          },
          validation: [
            {
              message: {
                id: 'test.field.validation.message',
                defaultMessage: 'Failed validation!',
                description: 'Test Field Validation Message Description'
              },
              validator: field('test.input').isEqualTo('helloooo')
            }
          ]
        }
      ],
      { 'test.input': 'not hello!' },
      {}
    )

    expect(errors).toMatchSnapshot()
  })

  it('should not return errors for field with custom validations if value passes validation', () => {
    const errors = getFieldErrors(
      [
        {
          id: 'test.input',
          type: FieldType.TEXT,
          required: true,
          label: {
            id: 'test.field.label',
            defaultMessage: 'Test Field',
            description: 'Test Field Description'
          },
          validation: [
            {
              message: {
                id: 'test.field.validation.message',
                defaultMessage: 'Failed validation!',
                description: 'Test Field Validation Message Description'
              },
              validator: field('test.input').isEqualTo('helloooo')
            }
          ]
        }
      ],
      { 'test.input': 'helloooo' },
      {}
    )

    expect(errors).toMatchSnapshot()
  })
})
