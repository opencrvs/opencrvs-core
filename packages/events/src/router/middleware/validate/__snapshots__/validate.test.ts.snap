// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`getFieldErrors() > should not return an error if a required field is provided 1`] = `[]`;

exports[`getFieldErrors() > should not return an error if a value for a conditionally hidden required field is not provided 1`] = `[]`;

exports[`getFieldErrors() > should not return an error if a value for a conditionally visible required field is provided 1`] = `[]`;

exports[`getFieldErrors() > should not return error if multiple fields with same id are configured, and value is provided 1`] = `[]`;

exports[`getFieldErrors() > should not return errors for field with custom validations if value passes validation 1`] = `[]`;

exports[`getFieldErrors() > should return an error if a required field is not provided 1`] = `
[
  {
    "id": "test.checkbox",
    "message": "Required for registration",
    "value": undefined,
  },
]
`;

exports[`getFieldErrors() > should return an error if a value for a conditionally hidden required field is provided 1`] = `
[
  {
    "id": "test.checkbox",
    "message": "Hidden or disabled field should not receive a value",
    "value": true,
  },
]
`;

exports[`getFieldErrors() > should return errors for field with custom validation if value does not pass validation 1`] = `
[
  {
    "id": "test.input",
    "message": "Failed validation!",
    "value": "not hello!",
  },
]
`;
